nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/flow/resource/hadoop/HDFSExternalResourceProvider.java

```diff
@@ -0,0 +1,262 @@
+        }
+
+        final String sourceDirectory = context.getProperties().get(SOURCE_DIRECTORY_PARAMETER);
+
+        if (sourceDirectory == null || sourceDirectory.isEmpty()) {
+            throw new IllegalArgumentException("Provider needs the source directory to be set");
+        }
+
+        this.sourceDirectory = new Path(sourceDirectory);
+        this.storageLocation = context.getProperties().get(STORAGE_LOCATION);
+
+        this.context = context;
+        this.initialized = true;
+    }
+
+    @Override
+    public Collection<ExternalResourceDescriptor> listResources() throws IOException {
+        if (!initialized) {
+            LOGGER.error("Provider is not initialized");
+        }
+
+        final HdfsResources hdfsResources = getHdfsResources();
+
+        try {
+
+            final FileStatus[] fileStatuses = hdfsResources.getUserGroupInformation()
+                    .doAs((PrivilegedExceptionAction<FileStatus[]>) () -> hdfsResources.getFileSystem().listStatus(sourceDirectory));
+
+            final List<ExternalResourceDescriptor> result = Arrays.stream(fileStatuses)
+                    .filter(fileStatus -> fileStatus.isFile())
+                    .map(HDFSExternalResourceProvider::convertStatusToDescriptor)
+                    .filter(context.getFilter().orElseGet(() -> flowResourceDescriptor -> true))
+                    .collect(Collectors.toList());
+
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("The following NARs were found: " + String.join(", ", result.stream().map(d -> d.getLocation()).collect(Collectors.toList())));
+            }
+
+            return result;
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Provider cannot list resources", e);
+        }
+    }
+
+    private static ExternalResourceDescriptor convertStatusToDescriptor(final FileStatus fileStatus) {
+        return new ImmutableExternalResourceDescriptor(fileStatus.getPath().getName(), fileStatus.getModificationTime());
+    }
+
+    @Override
+    public InputStream fetchExternalResource(final ExternalResourceDescriptor descriptor) throws IOException {
+        if (!initialized) {
+            LOGGER.error("Provider is not initialized");
+        }
+
+        final String location = descriptor.getLocation();
+        final Path path = getLocation(location);
+        final HdfsResources hdfsResources = getHdfsResources();
+
+        try {
+            if (hdfsResources.getUserGroupInformation().doAs((PrivilegedExceptionAction<Boolean>) () -> !hdfsResources.getFileSystem().exists(path))) {
+                throw new IOException("Provider cannot find " + location);
+            }
+
+            return hdfsResources.getUserGroupInformation()
+                    .doAs((PrivilegedExceptionAction<FSDataInputStream>) () -> hdfsResources.getFileSystem().open(path, BUFFER_SIZE_DEFAULT));
```

Comment:
Any reason not to combine this into a single call to `doAs`?
```
return hdfsResources.getUserGroupInformation().doAs((PrivilegedExceptionAction<FSDataInputStream>) () -> {
    if (!hdfsResources.getFileSystem().exists(path)) {
        throw new IOException("Cannot find file in HDFS at location " + location);
    }

    return hdfsResources.getFileSystem().open(path, BUFFER_SIZE_DEFAULT);
});
```
Or honestly, perhaps even just dropping the call to check if the file exists to begin with, as the call to `FileSystem.open()`should already throw an appropriate IOException